src/lib/pubkey/dl_group/dl_group.cpp

```diff
@@ -120,23 +129,24 @@ std::shared_ptr<DL_Group_Data> DL_Group::BER_decode_DL_group(const uint8_t data[
                                                              size_t data_len,
                                                              DL_Group_Format format,
                                                              DL_Group_Source source) {
-   BigInt p, q, g;
-
    BER_Decoder decoder(data, data_len);
    BER_Decoder ber = decoder.start_sequence();
 
    if(format == DL_Group_Format::ANSI_X9_57) {
+      BigInt p, q, g;
       ber.decode(p).decode(q).decode(g).verify_end();
+      return std::make_shared<DL_Group_Data>(p, q, g, source);
    } else if(format == DL_Group_Format::ANSI_X9_42) {
+      BigInt p, g, q;
       ber.decode(p).decode(g).decode(q).discard_remaining();
+      return std::make_shared<DL_Group_Data>(p, q, g, source);
    } else if(format == DL_Group_Format::PKCS_3) {
-      // q is left as zero
+      BigInt p, g;
       ber.decode(p).decode(g).discard_remaining();
+      return std::make_shared<DL_Group_Data>(p, g, source);
```

Comment:
Ugly, but it is parsing code, after all...

```C++
if(format == DL_Group_Format::ANSI_X9_57) {
   BigInt p, q, g;
   ber.decode(p).decode(q).decode(g).verify_end();
   return std::make_shared<DL_Group_Data>(std::move(p), std::move(q), std::move(g), source);
} else if(format == DL_Group_Format::ANSI_X9_42) {
   BigInt p, g, q;
   ber.decode(p).decode(g).decode(q).discard_remaining();
   return std::make_shared<DL_Group_Data>(std::move(p), std::move(q), std::move(g), source);
} else if(format == DL_Group_Format::PKCS_3) {
   BigInt p, g;
   ber.decode(p).decode(g).discard_remaining();
   return std::make_shared<DL_Group_Data>(std::move(p), std::move(g), source);
}
```

Reply:
These and others seem accurate but IMO not going to move the needle much since we have more heavy setup of Montgomery parameters etc which will dominate. Fine to do it but a little afield from the goal in thie PR so I'll pass for now.